third_party/src/main/java/com/jetbrains/lang/dart/dtd/DTDProcess.kt

```diff
@@ -0,0 +1,241 @@
+import com.jetbrains.lang.dart.sdk.DartSdk
+import com.jetbrains.lang.dart.sdk.DartSdkUtil
+import de.roderick.weberknecht.WebSocket
+import de.roderick.weberknecht.WebSocketEventHandler
+import de.roderick.weberknecht.WebSocketException
+import de.roderick.weberknecht.WebSocketMessage
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.atomic.AtomicInteger
+
+class DTDProcess {
+  companion object {
+    private val logger = logger<DTDProcess>()
+  }
+
+  private lateinit var osProcessHandler: KillableProcessHandler
+
+  private val consumerMap: MutableMap<String, DartToolingDaemonConsumer> = mutableMapOf()
+
+  private var processRunning = false
+
+  var uri: String? = null
+    private set
+  private var secret: String? = null
+
+  private val nextRequestId = AtomicInteger()
+
+  private lateinit var webSocket: WebSocket
+  var webSocketReady: Boolean = false
+    private set
+  private val servicesMap: MutableMap<String, DartToolingDaemonRequestHandler> = mutableMapOf()
+
+  private val eventDispatcher: EventDispatcher<DartToolingDaemonListener> =
+    EventDispatcher.create(DartToolingDaemonListener::class.java)
+
+  var listener: DTDProcessListener? = null
+    set
+  private fun connectToWebSocket(uri: String) {
+    try {
+      webSocket = WebSocket(URI(uri))
+      webSocket.eventHandler = object : WebSocketEventHandler {
+        override fun onClose() {
+          listener?.onWebSocketClose()
+        }
+
+        override fun onMessage(message: WebSocketMessage) {
+          val text = message.text
+          handleWebSocketMessage(text)
+          listener?.onWebSocketMessage(text)
+        }
+
+        override fun onOpen() {
+          webSocketReady = true
+          listener?.onWebSocketOpen()
+        }
+
+        override fun onPing() {}
+
+        override fun onPong() {}
+      }
+      webSocket.connect()
+    } catch (e: Exception) {
+      logger.error("Failed to connect to Dart Tooling Daemon, uri: $uri", e)
+    }
+  }
+
+  // TODO (pq): call from DartToolingDaemonService
+  fun addToolingDaemonListener(listener: DartToolingDaemonListener, parentDisposable: Disposable): Unit =
+    eventDispatcher.addListener(listener, parentDisposable)
+
+  private fun handleWebSocketMessage(text: String) {
+    logger.debug("<-- $text")
+
+    val json: JsonObject = try {
+      JsonParser.parseString(text) as JsonObject
+    } catch (e: Exception) {
+      logger.warn("Failed to parse message, error: ${e.message}, message: $text")
+      return
+    }
+
+    val method = json["method"]?.asString
+    val serviceConsumer = servicesMap[method]
+    if (method == "streamNotify") {
+      val params = json["params"].asJsonObject
+      val streamId = params["streamId"].asString
+      eventDispatcher.multicaster.received(streamId, json)
+    } else if (serviceConsumer != null) {
+      val params = json["params"].asJsonObject
```

Comment:
I made some updates to `DartToolingDaemonService` (https://github.com/flutter/dart-intellij-third-party/commit/9a70048f53c611f25bb4678201b10de10eb61b7f) that may conflict here. Mainly, I changed the IDs to be integers instead of strings and made params optional for service responses. Can you make sure those make it into `DTDProcess` before we refactor to use this in the service?